File: src/Term.ts
import { default as c } from "chalk";
import { terminal } from "terminal-kit";
import { Directory, File } from "./Tree";
import { Zip } from "./Zip";

export const [W, H] = process.stdout.getWindowSize();
let ct = 0;

export interface IFileInfo {
    path: string;
    size: number;
    outDir: string;
}

export class Term {
    static $: Directory;
    static $$: (Directory | File)[];
    static info: IFileInfo;
    static cursor: null | Directory | File = null;
    static frameStart = 0;
    static isExtracting = false;

    static async Init($: Directory, $$: (Directory | File)[], info: IFileInfo) {
        this.$ = $;
        this.$$ = $$;
        this.info = info;
        const stdin = process.stdin;
        stdin.setRawMode(true);
        stdin.resume();
        stdin.setEncoding("utf8");
        stdin.on("data", async (keyBuff) => {
            const key = keyBuff.toString("utf8");
            if (key == "\u0003") {
                process.exit();
            } // ctrl-c
            if (this.isExtracting) {
                return;
            }
            if (key == "\u001B\u005B\u0041" || key == "k") {
                //up
                this.CursorUp();
            }
            if (key == "\u001B\u005B\u0043" || key == "j") {
                //right
            }
            if (key == "\u001B\u005B\u0042") {
                //down
                this.CursorDown();
            }
            if (key == "\u001B\u005B\u0044") {
                //left
            }
            if (key === "\u000D") {
                if (this.cursor instanceof Directory) {
                    this.cursor.toggle();
                } else if (this.cursor instanceof File) {
                    this.cursor.toggle_mark();
                }
            }

            if (key === "w" && this.frameStart > 0) {
                this.frameStart--;
                this.CursorUp();
            }
            if (key === "s" && this.frameStart < this.$$.length) {
                this.frameStart++;
                this.CursorDown();
            }
            if (key === "m") {
                this.cursor?.toggle_mark();
            }
            if (key === "e") {
                this.Extract();
            }

            //console.log(Buffer.from(key).toString("hex"));

            await this.Draw();
        });
    }

    static async Extract() {
        const files = await this.$.files_to_extract();
        if (files.length === 0) {
            console.clear();
            console.log("No Files Selected");
            return;
        }
        this.isExtracting = true;
        await this.Draw();
        await Zip.Extract(this.info.path, this.info.outDir, files);
    }

    static CursorUp() {
        if (this.cursor !== null) {
            const showns = this.$$.filter((o) => o.shown);
            const ind = showns.indexOf(this.cursor);
            if (ind !== -1 && ind > 0) {
                this.cursor = showns[ind - 1];
            }
        }
    }

    static CursorDown() {
        if (this.cursor !== null) {
            const showns = this.$$.filter((o) => o.shown);
            const ind = showns.indexOf(this.cursor);
            if (ind !== -1 && ind < showns.length - 1) {
                this.cursor = showns[ind + 1];
            }
        }
    }

    static Write(s: string | Buffer) {
        process.stdout.write(s);
    }
    static async Header() {
        const line =
            "  Up/Down Move Cursor, w/s Move Frame, 'm' Mark 'e' Extract";
        let str = c.bgWhite.black(line + " ".repeat(W - line.length)) + "\n\n";
        str += "File: " + this.info.path + "\n";
        str += "Size: " + (this.info.size / 1024 / 1024).toFixed(2) + " MB\n";
        str += "\n";
        return str;
    }

    static async DrawTree() {
        // TODO: Line limiting and scrolling
        let str = "";
        let ct = 0;
        if (ct < this.frameStart) {
            str += "......\n";
        }
        for (const obj of this.$$) {
            if (obj.shown) {
                if (this.cursor === null) {
                    this.cursor = obj;
                }
                if (ct < this.frameStart) {
                    ct++;
                    continue;
                } else if (ct >= H - 7 + this.frameStart) {
                    str += "......\n";
                    break;
                } else {
                    ct++;
                }
                let line =
                    " ".repeat(obj.level) +
                    `${
                        obj instanceof File
                            ? `${obj.marked ? "[x] " : "[ ] "}`
                            : ""
                    }` +
                    obj.name;
                if (obj instanceof Directory) {
                    line += "/";
                }
                if (this.cursor === obj) {
                    str += c.bgWhite.black(
                        line + " ".repeat(W - line.length) + "\n"
                    );
                } else {
                    str += line + "\n";
                }
            }
        }
        return str;
    }

    static async Draw() {
        console.clear();
        if (this.isExtracting) {
            console.log("Extracting");
            return;
        }
        this.Write(await this.Header());
        this.Write(await this.DrawTree());
        terminal.moveTo(1, 1);
    }
}

File: src/term.rs
use crate::*;
use std::fs;
use std::io::{stdin, stdout, Read, Stdout, Write};
use std::path::Path;
use termion::color;
use termion::color::{Bg, Color, Fg, Reset};
use termion::cursor;
use termion::cursor::DetectCursorPos;
use termion::raw::{IntoRawMode, RawTerminal};

pub struct Term {
    _stdout: RawTerminal<Stdout>,
    root: Directory,
    out_dir: String,
    in_file: String,

    frame_start: usize,
    cursor: String,
    help_menu: bool,

    next_cursor: Option<String>,
    prev_cursor: Option<String>,
    will_down_move_frame: bool,
    will_up_move_frame: bool,
    is_cursor_dir: bool,
}

impl Term {
    pub fn new(root: Directory, in_file: String, out_dir: String) -> Self {
        Term {
            _stdout: stdout().into_raw_mode().unwrap(),
            cursor: root.get_name().to_owned(),
            root,
            frame_start: 0,
            next_cursor: None,
            prev_cursor: None,
            will_up_move_frame: false,
            will_down_move_frame: false,
            is_cursor_dir: true,
            out_dir,
            in_file,
            help_menu: false,
        }
    }
    pub fn draw(&mut self) {
        self.clear();
        let mut header: String = format!("  Help for '?'");
        let (w, h) = termion::terminal_size().unwrap();
        let space = " ".repeat((w as usize) - header.len());
        header = format!("{}{}", header, space);
        print!(
            "{}{}{}{}{}\r\n",
            Bg(color::White),
            Fg(color::Black),
            header,
            Bg(Reset),
            Fg(Reset)
        );

        print!("\r\n\r\n");

        if self.help_menu {
            println!("Up/Down or k/j to move\r\nEnter to open/close folders and mark files\r\n'm' to mark folders or files\r\n'e' to extract marked files and folders\r\nPress any key to close this menu\r");
            return;
        }

        let showns = self.root.shown_list();
        //calculate frame size
        let (_, ch) = self._stdout.cursor_pos().unwrap();
        let fh = (h - ch) - 3;
        if fh < 3 {
            self.clear();
            println!("Terminal Height Is Too Low. If Bugs Happen, Restart The Program\r");
            return;
        }
        //re-adjust frame start
        //if there is a space in the frame
        //adjsut frame
        if showns.len() - self.frame_start < fh as usize {
            self.frame_start = isize::max(0, (showns.len() as isize) - (fh as isize)) as usize;
            // TODO: cursor should be visible but make sure of that
        }
        if self.frame_start != 0 {
            println!("...\r");
        }

        let last_pos = usize::min(showns.len(), self.frame_start + fh as usize);
        for k in self.frame_start..last_pos {
            let el = &showns[k];
            let is_cursor = el.get_full_path() == self.cursor;
            if is_cursor {
                //set all flags
                if k == self.frame_start {
                    if self.frame_start != 0 {
                        panic!("Frame Start is not zero and cursor is at top");
                    }
                    self.will_up_move_frame = false;
                    self.prev_cursor = None;
                } else if k == self.frame_start + 1 {
                    //means cursor is 1 element below of first element
                    self.will_up_move_frame = self.frame_start > 0;
                    self.prev_cursor = Some((&showns[k - 1]).get_full_path().to_owned());
                } else {
                    self.will_up_move_frame = false;
                    self.prev_cursor = Some((&showns[k - 1]).get_full_path().to_owned());
                }

                if k == last_pos - 1 {
                    // above the last element
                    if k != showns.len() - 1 {
                        //if we are at bottom and we are not at last element,
                        //there is a problem in our logic
                        panic!("We are at bottom but there are more elements");
                    }
                    self.will_down_move_frame = false;
                    self.next_cursor = None;
                } else if k == last_pos - 2 {
                    //last element
                    self.will_down_move_frame = self.frame_start + (fh as usize) < showns.len();
                    self.next_cursor = Some((&showns[k + 1]).get_full_path().to_owned());
                } else {
                    self.will_down_move_frame = false;
                    self.next_cursor = Some((&showns[k + 1]).get_full_path().to_owned());
                }
            }
            if is_cursor {
                print!("{}{}", Bg(color::White), Fg(color::Black));
            } else {
                print!("{}{}", Bg(color::Reset), Fg(color::Reset));
            };
            match el {
                Node::File(f) => {
                    print!(
                        "{}[{}] {}",
                        "  ".repeat(f.get_level() as usize),
                        if f.is_marked() { "x" } else { " " },
                        f.get_name()
                    );
                    if is_cursor {
                        self.is_cursor_dir = false;
                    }
                }
                Node::Directory(d) => {
                    print!("{}{}/", "  ".repeat(d.get_level() as usize), d.get_name());
                    if is_cursor {
                        self.is_cursor_dir = true;
                    }
                }
            }
            print!("{}{}", Bg(color::Reset), Fg(color::Reset));
            print!("\r\n");
        }

        // k == last_pos - 1 should work too
        if self.frame_start + (fh as usize) < showns.len() {
            println!("...\r");
        }

        print!("{}", cursor::Hide);
        stdout().flush().unwrap();
    }
    fn clear(&mut self) {
        print!("{esc}[2J{esc}[1;1H", esc = 27 as char);
    }

    fn write_list_file(&self) -> String {
        let list = self.root.marked_list();
        let path_string = if Path::new("/tmp").exists() {
            format!("/tmp/.7zb-list-test.tmp")
        } else {
            format!("./.7zb-list-test.tmp")
        };
        let path = Path::new(&path_string);
        let mut file = fs::File::create(path).expect("Temporary list file creation error");
        for n in list {
            file.write(n.get_full_path().as_bytes())
                .expect("Temporary list write error");
            file.write("\n".as_bytes())
                .expect("Temporary list write error2");
        }
        path_string
    }

    pub fn ui_loop(&mut self) -> Option<String> {
        self.draw();
        //println!("{:#?}", self.next_cursor);
        for keycode in stdin().bytes() {
            let keycode = keycode.unwrap();
            self.help_menu = false;

            if keycode == 3 || keycode == 113 {
                print!("{}", cursor::Show);
                return None;
            }
            if keycode == 65 || keycode == 'k' as u8 {
                //up
                if self.will_up_move_frame {
                    self.frame_start -= 1;
                }
                if let Some(c) = &self.prev_cursor {
                    self.cursor = c.clone();
                }
            } else if keycode == 66 || keycode == 'j' as u8 {
                //down
                if self.will_down_move_frame {
                    self.frame_start += 1;
                }
                if let Some(c) = &self.next_cursor {
                    self.cursor = c.clone();
                }
            } else if keycode == 67 {
                //right
            } else if keycode == 68 {
                //left
            } else if keycode == 13 {
                //enter
                if self.is_cursor_dir {
                    self.root.toggle_open(&self.cursor);
                } else {
                    self.root.toggle_marked_file(&self.cursor);
                }
            } else if keycode == 'm' as u8 {
                //m
                if self.is_cursor_dir {
                    self.root.toggle_marked(&self.cursor);
                } else {
                    self.root.toggle_marked_file(&self.cursor);
                }
            } else if keycode == 'e' as u8 {
                self.clear();
                print!("{}", cursor::Show);
                self._stdout.flush().unwrap();
                let list_path = self.write_list_file();
                return Some(list_path);
            } else if keycode == '?' as u8 {
                self.help_menu = true;
            }
            self.draw();
            //println!("{}\r", keycode);
        }
        print!("{}", cursor::Show);
        None
    }
}
